### PR TITLE
feat(ColorPicker): add onClear

### PR DESCRIPTION
**Pull request by linxianxi**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [x] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

https://github.com/ant-design/ant-design/pull/42406

### 💡 Background and solution

应该让用户自己控制清除时是否关闭，用 open 受控配合 onClear 实现，或者加一个类似 closePopupWhenClear 属性。

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     add onClear and dont't close panel when click      |
| 🇨🇳 Chinese |     添加 onClear，清除时不自动关闭弹窗      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at e45c232</samp>

Added a new `onClear` prop to the `ColorPicker` component to customize the behavior when clearing the color. Refactored the color picker panel props and the popup closing and color clearing logic. Updated the documentation and the test case accordingly.

### <samp>🤖 Generated by Copilot at e45c232</samp>

*  Add `onClear` prop to `ColorPickerProps` interface and `ColorPicker` component to allow custom callback function when clearing the color ([link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR47), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR75), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fL118-R122), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-819859f3d30fc8ec12a8ef1b55ac1d75b4b1d600eb1cee989efd5af57255b560L39-R40), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-819859f3d30fc8ec12a8ef1b55ac1d75b4b1d600eb1cee989efd5af57255b560L48-R52), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-9d1ff106a032e484c8f982279c4e1295d23f69904a26e020e99310c446e93249L40-R41), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-9d1ff106a032e484c8f982279c4e1295d23f69904a26e020e99310c446e93249L49-R53))
*  Refactor `handleClear` function in `ColorPicker` component to set `colorCleared` state and call `onClear` prop function without passing boolean parameter ([link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fL118-R122))
*  Remove `useEffect` hook and import from `ColorPicker` component as popup closing logic is moved to `handleClear` function ([link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fL8-R8), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fL143-L148))
*  Change `onClear` prop in `ColorPickerPanelProps` interface and `ColorPickerPanel` component to a function without parameters to match `ColorPickerProps` and `handleClear` function ([link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-3abf1a34a643a7087dfbaf3039447778574de2b61e627f4c1218297fb961c70fL14-R14), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-3abf1a34a643a7087dfbaf3039447778574de2b61e627f4c1218297fb961c70fL29-R29))
*  Move `onFormatChange` prop from `ColorPickerProps` to `ColorPickerPanelProps` as it is more relevant for customizing color format ([link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-819859f3d30fc8ec12a8ef1b55ac1d75b4b1d600eb1cee989efd5af57255b560L39-R40), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-819859f3d30fc8ec12a8ef1b55ac1d75b4b1d600eb1cee989efd5af57255b560L48-R52), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-9d1ff106a032e484c8f982279c4e1295d23f69904a26e020e99310c446e93249L40-R41), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-9d1ff106a032e484c8f982279c4e1295d23f69904a26e020e99310c446e93249L49-R53))
*  Move `onChange` and `onOpenChange` props from `ColorPickerProps` to `ColorPickerPanelProps` as they are more relevant for customizing color picker panel behavior ([link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-819859f3d30fc8ec12a8ef1b55ac1d75b4b1d600eb1cee989efd5af57255b560L48-R52), [link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-9d1ff106a032e484c8f982279c4e1295d23f69904a26e020e99310c446e93249L49-R53))
*  Remove assertion that popup should be hidden after clearing color from test case in `components/color-picker/__tests__/index.test.tsx` as it is handled by `handleClear` function ([link](https://github.com/ant-design/ant-design/pull/42634/files?diff=unified&w=0#diff-aa06460351e8630f305cf6509e8137984e0f167ab2560da22874033189234fb4L86-R89))

### CODE REVIEW

**Comment by RedJue** [comment]
补个 `onClear` 的单测，`allowClear` demo 可以完善一下，加个 `onClear` 控制 panel 关闭的例子。

**Comment by RedJue** [comment]
单测挂了，本地调试一下。

**Comment by linxianxi** [comment]
@RedJue 这个单侧不知道怎么不行，你看看

**Comment by linxianxi** [comment]
还有一个点击清除，会一直触发 onChange 和 onClear，这是故意为之还是 bug？已经是 clear 状态再点清除不是应该什么都不发生吗

**Comment by RedJue** [comment]
> 还有个问题，一直点击清除，会一直触发 onChange 和 onClear，0 和 100 会反复切换，这是故意为之还是 bug？已经是 clear 状态再点清除不是应该什么都不发生吗

是 `bug` 如果清空状态下不应该被点击，这个可以单独提个 fix修复。

**Comment by kiner-tang** [comment]
> > 还有个问题，一直点击清除，会一直触发 onChange 和 onClear，0 和 100 会反复切换，这是故意为之还是 bug？已经是 clear 状态再点清除不是应该什么都不发生吗
> 
> 是 `bug` 如果清空状态下不应该被点击，这个可以单独提个 fix修复。

这个可以跟 清空后不隐藏面板 的问题一起修复，然后提 pr 到 master

File: components/color-picker/__tests__/index.test.tsx [hunk 1]
```diff
@@ -79,16 +79,16 @@ describe('ColorPicker', () => {
...
   });
 
-  it('Should allowClear work', async () => {
-    const { container } = render(<ColorPicker allowClear />);
+  it('Should allowClear and onClear work', async () => {
+    const onClear = jest.fn();
+    const { container } = render(<ColorPicker allowClear onClear={onClear} />);
     fireEvent.click(container.querySelector('.ant-color-picker-trigger')!);
     await waitFakeTimer();
-    expect(container.querySelector('.ant-popover-hidden')).toBeFalsy();
```
Comment (RedJue) [inline components/color-picker/__tests__/index.test.tsx:86]: 这里不需要去掉，点击之后 弹窗应该弹出来。

Comment (linxianxi) [reply components/color-picker/__tests__/index.test.tsx:86]: 加上了之后，最后那行还是 0%，得不到 100%

**Comment by RedJue** [comment]
> @RedJue 这个单测不知道怎么不行，你看看

猜测可能单测里行为和现在的预期不一样，尝试删掉这几行 https://github.com/ant-design/ant-design/pull/42634/files#diff-aa06460351e8630f305cf6509e8137984e0f167ab2560da22874033189234fb4L99-L103 试试

File: components/color-picker/__tests__/index.test.tsx (hunk 1 above)
Comment (RedJue) [reply components/color-picker/__tests__/index.test.tsx:86]: 这个和单测结果错误不相关。

Comment (linxianxi) [reply components/color-picker/__tests__/index.test.tsx:86]: 我知道不一样，删掉那几行是可以了，但是这几行看上去感觉没问题。

Comment (RedJue) [reply components/color-picker/__tests__/index.test.tsx:86]: 有问题，现在是点击清空不关闭弹窗，删掉那部分逻辑是按之前点击清空关闭弹窗来走的，按现在逻辑再点击就关闭了，自然触发就有问题。

**Comment by linxianxi** [comment]
https://github.com/ant-design/ant-design/pull/42643  等这个合到 master，这里的 pr 再更新

File: components/color-picker/__tests__/index.test.tsx (hunk 1 above)
Comment (linxianxi) [reply components/color-picker/__tests__/index.test.tsx:86]: 这几行意思是点击一下方块，获取值是 0，应该没问题吧，跟是否关闭弹窗有关系么。删了删了，等另一个 pr 合并再更新这里的。

File: components/color-picker/demo/allowClear.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 ## zh-CN
 
-清除已选择的颜色。
+清除已选择的颜色。可以使用受控 open 和 onClear 控制清除时关闭弹窗。
```
Comment (RedJue) [inline components/color-picker/demo/allowClear.md:3]: demo 分两个吧，一个最基础的用法，一个配合 `onClear ` 实现弹窗关闭。

Comment (linxianxi) [reply components/color-picker/demo/allowClear.md:3]: 其实我觉得就只要最基础的用法就好了，配合 onClear 关闭没有必要。

Comment (RedJue) [reply components/color-picker/demo/allowClear.md]: 把用户当傻子就对了，能给就给。

Comment (RedJue) [reply components/color-picker/demo/allowClear.md]: 不加应该问题也不大，先这样吧。

**Review by RedJue — APPROVED** [review]

**Pull request merged**